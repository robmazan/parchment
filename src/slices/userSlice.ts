import { createSlice, PayloadAction } from "@reduxjs/redux-toolkit";

export type User = {
  username: string;
  firstname: string;
  lastname: string;
  roles: string[];
};

export const slice = createSlice({
  name: "user",
  initialState: (null as unknown) as User,
  reducers: {
    receiveUser: (state: User, action: PayloadAction<User>): User => {
      return action.payload;
    }
  }
});

export default slice;

export const getUserFullName = (state?: User): string => {
  return state ? [state.firstname, state.lastname].join(" ").trim() : "";
};

export const getIsLoggedIn = (state?: User): boolean => {
  if (!state) {
    return false;
  }
  return state !== null;
};
