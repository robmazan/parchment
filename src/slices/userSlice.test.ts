import * as userSlice from "./userSlice";
import { AnyAction } from "redux";

const sampleData: userSlice.User = require("./__mocks__/sampleUser.json");

describe("user slice", () => {
  describe("receiveUser action", () => {
    it("creates an action to receive user", () => {
      const actionCreator = userSlice.slice.actions.receiveUser;
      expect(actionCreator(sampleData)).toMatchSnapshot();
    });
  });

  describe("combined reducer", () => {
    it("returns the expected default initial state", () => {
      const initialState = userSlice.slice.reducer(undefined, {} as AnyAction);
      expect(initialState).toEqual(null);
    });

    it("handles receiveUser action", () => {
      const reducer = userSlice.slice.reducer;
      const actionCreator = userSlice.slice.actions.receiveUser;
      expect(reducer(undefined, actionCreator(sampleData))).toEqual(sampleData);
    });
  });

  describe("getUserFullName selector", () => {
    it("returns user's fullname", () => {
      const received = userSlice.getUserFullName(sampleData);
      const expected = [sampleData?.firstname, sampleData?.lastname]
        .join(" ")
        .trim();
      expect(received).toEqual(expected);
    });

    it("returns empty string if user is undefined", () => {
      const received = userSlice.getUserFullName(undefined);
      const expected = "";
      expect(received).toEqual(expected);
    });
  });

  describe("getIsLoggedIn selector", () => {
    it("returns false if user is undefined", () => {
      const received = userSlice.getIsLoggedIn(undefined);
      expect(received).toBe(false);
    });

    it("returns true if user is defined", () => {
      const received = userSlice.getIsLoggedIn(sampleData);
      expect(received).toBe(true);
    });
  });
});
