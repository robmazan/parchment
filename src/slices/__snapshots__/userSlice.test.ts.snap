// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user slice receiveUser action creates an action to receive user 1`] = `
Object {
  "payload": Object {
    "firstname": " Uther ",
    "lastname": " User ",
    "roles": Array [
      "user",
    ],
    "username": "Uther User",
  },
  "type": "user/receiveUser",
}
`;
